content/source/docs/extend/resources.html.md

```diff
@@ -0,0 +1,368 @@
+            "name": {
+                Type:     schema.TypeString,
+                Required: true,
+            },
+        },
+        Timeouts: &schema.ResourceTimeout{
+            Create: schema.DefaultTimeout(45 * time.Minute),
+        },
+    }
+}
+```
+
+In the above example we see the usage of the timeouts in the schema being configured for what is deemed the appropriate amount of time for the `Create` function. `Read`, `Update`, and `Delete` are also configurable as well as a `Default`. These configured timeouts can be fetched later in the CRUD functions from the passed in `*schema.ResourceData`.
+
+### Retry
+
+A common case for requiring retries or polling is when the backend infrastructure being provisioned is designed to be asynchronous, requiring the developer to repeatedly check the status of the resource. The retry helper takes a timeout and a function that is retried repeatedly. The timeout can be retrieved from the `*schema.ResourceData` struct, using the `Timeout` method, passing in the appropriate timeout key (`scheme.TimeoutCreate`). The retry function provided should return either a `resource.NonRetryableError` for unexpected errors or states, otherwise continue to retry with a `resource.RetryableError`. In the context of a `CREATE` function, once the backend responds with the desired state, finish the function with a `resource.NonRetryableError` wrapping the `READ` function (anything that goes wrong in there is considered unexpected).
+
+```go
+func resourceExampleInstanceCreate(d *schema.ResourceData, meta interface{}) error {
+    name := d.Get("name").(string)
+    client := meta.(*ExampleClient)
+    resp, err := client.CreateInstance(name)
+
+    if err != nil {
+        return fmt.Errorf("Error creating instance: %s", err)
+    }
+
+    return resource.Retry(d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
+        resp, err := client.DescribeInstance(name)
+
+        if err != nil {
+            return resource.NonRetryableError(fmt.Errorf("Error describing instance: %s", err))
+        }
+
+        if resp.Status != "CREATED" {
+            return resource.RetryableError(fmt.Errorf("Expected instance to be created but was in state %s", resp.Status))
+        }
+
+        return resource.NonRetryableError(resourceExampleInstanceRead(d, meta))
+    })
+}
+```
+
+### StateChangeConf
+
+`resource.Retry` is useful for simple scenarios, but sometimes handling an APIs latency or eventual consistency requires more fine tuning. `resource.Retry` is in fact a wrapper for a another helper: `resource.StateChangeConf`. Use `resource.StateChangeConf` when your resource has several pending states, the retry timing and backoff requires fine-tuning, or you want to ensure a minimum number of occurrences of a target state is reached (this is very common when dealing with eventually consistent APIs, where a response can reply back with an old state between calls before becoming consistent).
```

Comment:
I think we should explain the main difference between the two:

`resource.Retry` is useful for retrying on errors or more generally when dealing with **two** states (error / success).
`resource.StateChangeConf` should be preferred when the resource can go through **multiple** states as it provides more fine grained control and set expectations around what state can the resource be in.

It doesn't have to be word by word, but this is how I understand the two differ, and how I would decide which one to use.

Reply:
👍